Create a hydra-explorer executable that can track all heads on-chain

ffakenz opened the pull request:
<!-- Describe your change here -->

# What

Create a hydra-explorer that can track all Babbage heads on-chain in a devnet network.
> no multi-version support needed (for now).

Then expose a REST-API to query the "List of all heads and their current state (Initializing, Open, Closed, Final) on network"

# How

* [x] Write an e2e test for hydra-explorer to drive the development; opening a single head and using the process std out to assert.
* [x] Create hydra-explorer submodule and base process.
* [x] Reuse the hydra-chain-observer library code to collect the observations
* [x] Write an e2e test for hydra-explorer driving the development; opening multiple heads and using a REST API to assert.
* [x] Store observations in-memory and maintain a dictionary of heads and states.
* [x] Create a backend http-server for hydra-explorer that binds port 9090.
* [x] Expose an api to query the "List of all heads and their current state (Initializing, Open, Closed, Final) on network" observed
- GET /heads

---

<!-- Consider each and tick it off one way or the other -->
* [ ] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [ ] No new TODOs introduced or explained herafter

ffakenz committed:
Add base e2e test for hydra-exporer

To assert observations over stdout.

ffakenz committed:
Create base explorer project

Include changes to cabal and nix config files.

ffakenz committed:
Reuse the hydra-chain-observer library code to collect the observations

ffakenz committed:
Fix after rebase: remove pparams arg from withHydraNode process

ffakenz committed:
Add GET /heads e2e spec

ffakenz committed:
Draft http server for explorer

ffakenz committed:
Introduce ObserverHandler as callback to chain-observer

As a mean for clients to perform actions on every observation (like: keeping state)

ffakenz committed:
Simplify endpoint

ffakenz committed:
Add e2e spec to explore multiple heads in the past

ffakenz committed:
Add check process has not die to hydra-explorer in e2e specs

ffakenz committed:
Fix minor typo on observations

ch1bo committed:
Make hydra-cluster depend on hydra-explorer to run tests

This avoids need for installing the binary

ch1bo committed:
Add point to RollForward to see slots pass

ch1bo committed:
Refactor explorer e2e test slightly

ffakenz committed:
Start explorer from genesis

ffakenz committed:
Enhance e2e spec over api

ch1bo committed:
Draft some plumbing in explorer HTTP server

ffakenz committed:
Make use of ObserverHandler during roll forward

Simplify ObserverHandler and ExplorerState types.

ch1bo committed:
Fix GetHeadIds read model

We must access the TVar in the actual GetHeadIds IO action instead only
when setting it up.

ffakenz committed:
Draft openapi

ffakenz committed:
Introduce ipc server on chain-observer

Remove the need to pass a handler to the main fn.

ffakenz committed:
Revert "Introduce ipc server on chain-observer"

This reverts commit 1f21d9556db9b258d46c8b7164aa722a06a6c2ee.

ffakenz committed:
Enhance explorer state to be an aggregation of head observations

ffakenz committed:
Aggregate observations into explorer state

v0d1ch committed:
Fix assertion for the HeadId equality

ffakenz committed:
Add api server spec to validate openapi schema against server responses

Also added json content-type header to the GET /heads endpoint.

ffakenz committed:
Fix match openapi schema spec

Do not use cardanonical refs due to openapi schema validation does not support
to resolve external refs using $ref.

ffakenz committed:
Ehance e2e spec for explorer

So it showcases the explorer keeping track of multiple heads.

Also make head state aggregation to append the new heads states
to the tail of the explorer state.

v0d1ch commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+{-# LANGUAGE DeriveAnyClass #-}
+-- withCreateProcess interface is annoying
```
I think this is some leftover comment.

ffakenz replied:
good catch!

v0d1ch commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+            aliceChainConfig <- chainConfigFor Alice tmpDir nodeSocket hydraScriptsTxId [] cperiod
+            withHydraNode hydraTracer aliceChainConfig tmpDir 1 aliceSk [] [1] $ \hydraNode -> do
+              withHydraExplorer cardanoNode $ \explorer -> do
+                seedFromFaucet_ cardanoNode aliceCardanoVk 100_000_000 (contramap FromFaucet tracer)
```
Maybe move this line a bit up so it doesn't introduce any noise when looking at the more important test code.

v0d1ch commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+                waitFor hydraTracer 5 [hydraNode] $
+                  output "HeadIsOpen" ["utxo" .= object mempty, "headId" .= headId]
+
+                headExplorerSees explorer "HeadCommitTx" headId
```
I would perhaps move this line right below `requestCommitTx` to make sure the exact flow is observed in sequence.

ffakenz replied:
this test got unnecessary and will be removed since we only care about the http api :)

v0d1ch commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+            aliceHeadId <- withHydraNode hydraTracer aliceChainConfig tmpDir 1 aliceSk [] [1] initHead
+
+            (bobCardanoVk, _bobCardanoSk) <- keysFor Bob
+            seedFromFaucet_ cardanoNode bobCardanoVk 100_000_000 (contramap FromFaucet tracer)
```
The seed ada amount of 100 is probably a bit too much especially since we only init and finish the test.

v0d1ch commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+            withHydraExplorer cardanoNode $ \explorer -> do
+              (aliceCardanoVk, _aliceCardanoSk) <- keysFor Alice
+              aliceChainConfig <- chainConfigFor Alice tmpDir nodeSocket hydraScriptsTxId [] cperiod
+              aliceHeadId <- withHydraNode hydraTracer aliceChainConfig tmpDir 1 aliceSk [] [1] $ \hydraNode -> do
```
No big deal but maybe we want to set the hydra keys here also (second to last argument) to `withHydraNode`. This is also true for the test above.

v0d1ch replied:
Actually not needed since you run just a single party heads here.

ch1bo commented on hydra-explorer/hydra-explorer.cabal:
```diff
@@ -0,0 +1,99 @@
+  build-depends:
+    , aeson
+    , base
+    , bytestring
```
Should: remove unused packages

Some of these are not required (see compiler warning)

ch1bo commented on hydra-explorer/README.md:
````diff
@@ -0,0 +1,12 @@
+``` shell
+hydra-explorer \
+  --node-socket testnets/preprod/node.socket \
+  --testnet-magic 1 \
````
```suggestion
  --testnet-magic 1
```

ch1bo replied:
This is assuming we run a preprod node (maybe include that command here as well)

ch1bo replied:
Should: drop the non-json logs when starting the `hydra-explorer` (as we have all info in json outputs)

v0d1ch commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+
+                pure bobHeadId
+
+              manager <- HTTPClient.newTlsManager
```
I don't think you need to use tls here, we don't hit `https` endpoint. Also there is a  `parseRequestThrow` already in scope so you might want to use `Network.HTTP.Simple` and then we don't need extra dependencies added in `hydra-cluster.cabal` file.

ffakenz replied:
fyi: this does requires me to add `http-conduit` as dependency

v0d1ch commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+
+newtype HydraExplorerHandle = HydraExplorerHandle {awaitNext :: IO Value}
+
+data HydraExplorerLog
```
There might be a better place for these types.

v0d1ch replied:
It's a bit hard to move them since they depend on hydra-cluster...Let's leave them as is.

v0d1ch commented on hydra-explorer/README.md:
```diff
@@ -0,0 +1,12 @@
+A small executable which connects to a chain like the `hydra-node`, but puts any
+observations as traces onto `stdout`.
+
+To run, pass a `--node-socket`, corresponding network id. For example:
```
```suggestion
To run, pass a `--node-socket` and the corresponding network id. For example:
```

v0d1ch commented on hydra-explorer/src/Hydra/Explorer.hs:
```diff
@@ -0,0 +1,114 @@
+    let getHeads = readModelGetHeadIds explorerState
+    args <- getArgs
+    race
+      -- FIXME: this is going to be problematic on mainnet.
```
Can we fix the fixme? :) Or add a red bin item once this is merged.

ffakenz committed:
Fix chain-observer spec as observeTx is now returning a head observation

v0d1ch commented on hydra-explorer/src/Hydra/Explorer.hs:
```diff
@@ -0,0 +1,114 @@
+    ("HEAD", _) -> send $ responseLBS status200 corsHeaders ""
+    ("GET", []) -> send $ handleFile "index.html"
+    ("GET", ["heads"]) -> handleGetHeads getHeads req send
+    -- FIXME: do proper file serving, this is dangerous
```
Ok, I guess there will be continuation of work for head-explorer where we will resolve these fixmes.

ffakenz replied:
I've removed all endpoints serving files as they are not needed for the moment.

v0d1ch commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -0,0 +1,197 @@
+
+import Hydra.Prelude
+
+-- XXX: Depending on hydra-node will be problematic to support versions
```
This comment seems displaced.

v0d1ch commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -0,0 +1,197 @@
+instance Arbitrary HeadMember where
+  arbitrary = genericArbitrary
+
+data HeadStatus
```
A bit annoying that we have similar but not exacly the same type in the `ServerOutput.HeadStatus`

ch1bo commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+
+            withHydraExplorer cardanoNode $ \explorer -> do
+              headExplorerSees explorer "HeadInitTx" aliceHeadId
+              headExplorerSees explorer "HeadInitTx" bobHeadId
```
Should: assert REST API behavior rather than std out logs here.

ffakenz committed:
Remove unused packages

ffakenz committed:
Minor fix to readme run cmd

ffakenz committed:
Remove prints before server starts as traces already log this info

ch1bo commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+                    }
+              response <- HTTPClient.httpLbs request manager
+              HTTPClient.responseStatus response `shouldBe` status200
+              allHeads <- unsafeDecodeJson . toStrict $ HTTPClient.responseBody response
```
Could: use the HTTP.Simple modules to make this a lot shorter

https://hackage.haskell.org/package/http-conduit-2.3.8.3/docs/Network-HTTP-Simple.html#v:httpJSON

ch1bo commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+                HTTPClient.parseRequest url <&> \request ->
+                  request
+                    { HTTPClient.method = "GET"
+                    , HTTPClient.requestHeaders = [("Accept", "application/json")]
```
Could use [addRequestHeader](https://hackage.haskell.org/package/http-conduit-2.3.8.3/docs/Network-HTTP-Simple.html#v:addRequestHeader) or maybe `Accept: application/json` is already the default?

ch1bo commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+                pure bobHeadId
+
+              manager <- HTTPClient.newTlsManager
+              let url = "http://127.0.0.1:9090/heads"
```
```suggestion
              response <-
                parseRequestThrow "http://127.0.0.1:9090/heads"
                  <&> setRequestHeader "Accept" "application/json"
                  >>= httpJSON
```

ch1bo commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+  align _ [] = []
+  align n (h : q) = h : fmap (T.replicate n " " <>) q
+
+newtype HydraExplorerHandle = HydraExplorerHandle {awaitNext :: IO Value}
```
Could: use this handle as an easy access to the API (= a client), e.g.

```suggestion
newtype HydraExplorerHandle = HydraExplorerHandle {getHeads :: IO [Value]}
```

where `getHeads` does the request building, sending and response parsing.

Note that I suggest to use `Value` to avoid relying on the JSON serialization of the server and rather force the integration test to say what fields are expected (keepin the assertions to a minimum)

ch1bo commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+              encode aliceHeadId `shouldBe` encode aliceIdBS
+              aliceStatus `shouldBe` Initializing
+              encode bobHeadId `shouldBe` encode bobIdBS
+              bobStatus `shouldBe` Aborted
```
Must: not use the server types to implement the assertion. e.g something like this

```suggestion
              length allHeads `shouldBe` 2
              allHeads ^. at 0 . key "headId" `shouldBe` aliceIdBS
              allHeads ^. at 0 . key "status" `shouldBe` "Initializing"
              allHeads ^. at 1 . key "headId" `shouldBe` bobIdBS
              allHeads ^. at 1 . key "status" `shouldBe` "Aborted"
```

ch1bo commented on hydra-cluster/test/Test/ChainObserverSpec.hs:
```diff
@@ -112,10 +112,6 @@ data ChainObserverLog
-  -- to the file path opened (e.g. the cardano-node log file) in the test
-  -- failure output. Print the exception here to have some debuggability at
-  -- least.
   handle (\(e :: IOException) -> print e >> throwIO e) $
```
```suggestion
```

ch1bo commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,239 @@
+-- | Starts a 'hydra-explorer' on some Cardano network.
+withHydraExplorer :: RunningNode -> (HydraExplorerHandle -> IO ()) -> IO ()
+withHydraExplorer cardanoNode action =
+  handle (\(e :: IOException) -> print e >> throwIO e) $
```
```suggestion
```

v0d1ch commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -0,0 +1,197 @@
+    -- REVIEW: this should never happen; how should we deal with this scenario?
+    Nothing -> explorerState
+
+aggregateCloseObservation :: CloseObservation -> ExplorerState -> ExplorerState
```
all these aggregate functions are similar, perhaps we can write more general function for aggregation?

ch1bo commented on hydra-explorer/test/Hydra/Explorer/API/HTTPServerSpec.hs:
```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Hydra.Explorer.API.HTTPServerSpec where
```
Should: match the module this tests

There is no `Hydra.Explorer.API.HTTPServer` module. So maybe create it or rename this test to where the actual routes are implemented (`Hydra.ExplorerSpec`)

ch1bo commented on hydra-explorer/test/Hydra/Explorer/API/HTTPServerSpec.hs:
```diff
@@ -0,0 +1,76 @@
+                                    , ("Access-Control-Allow-Methods", "*")
+                                    , ("Access-Control-Allow-Headers", "*")
+                                    , ("Content-Type", "application/json")
+                                    ]
```
Should: only assert what we care about, i.e. that it's `Content-type: application/json`

ch1bo commented on hydra-explorer/test/Hydra/Explorer/API/HTTPServerSpec.hs:
```diff
@@ -0,0 +1,76 @@
+                                    ]
+              let maybeValue = Aeson.decode simpleBody :: Maybe Aeson.Value
+              case maybeValue of
+                Nothing -> liftIO . failure $ "Failed to decode body into json value"
```
Should: use `eitherDecode` to print why the responded bytes where not a proper JSON value

v0d1ch reviewed (approved):
Nice work, I only have some superficial comments.

ch1bo commented on hydra-explorer/src/Hydra/Explorer.hs:
```diff
@@ -0,0 +1,114 @@
+    ("GET", []) -> send $ handleFile "index.html"
+    ("GET", ["heads"]) -> handleGetHeads getHeads req send
+    -- FIXME: do proper file serving, this is dangerous
+    ("GET", path) -> send $ handleFile $ toString $ mconcat $ List.intersperse "/" ("." : path)
```
```suggestion
```

ch1bo commented on hydra-explorer/src/Hydra/Explorer.hs:
```diff
@@ -0,0 +1,114 @@
+      }
+  case (requestMethod req, pathInfo req) of
+    ("HEAD", _) -> send $ responseLBS status200 corsHeaders ""
+    ("GET", []) -> send $ handleFile "index.html"
```
```suggestion
```

ch1bo commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -0,0 +1,197 @@
+        then newHeadState : tailStates
+        else currentHeadState : replaceHeadState newHeadState tailStates
+
+aggregateHeadObservations :: [HeadObservation] -> ExplorerState -> ExplorerState
```
Should: (property) test this function

Ideas for properties:
* Given any observations, if a head is in the state, it is still in the state.
* Given any observations, the resulting list of heads (their ids) is a prefix of the original list of heads (their ids). (= generalization of the first bullet)
* Check the behavior on re-ordered the observations?

Given these we can express our expectation about the explorer behavior in terms of potentially re-ordered observations, or maybe it relies on things being in order (which would be perfectly fine).

ch1bo replied:
Must: fulfill property

> Any head observation (of some head id) must yield an entry of that head id

Because, this ensures that the explorer always _at least_ knows about the existence of a head. Even if we only observe a part of the life cycle of some head.

ch1bo replied:
```
forAll arbitrary $ \observations ->
   aggregateHeadObservations observations [] =/= []
```

ch1bo commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -0,0 +1,197 @@
+
+data HeadState = HeadState
+  { headId :: HeadId
+  , seedTxIn :: TxIn
```
```suggestion
  , seedTxIn :: Maybe TxIn
```

or even better:

```
data Observed a = Unknown | Seen a
```

and

```
 seedTxIn :: Observed TxIn
```

ch1bo replied:
Including an appropriated schema

ch1bo commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -0,0 +1,197 @@
+data HeadMember = HeadMember
+  { party :: Party
+  , onChainId :: OnChainId
+  , commits :: [PartyCommit]
```
Should: not re-invent the wheel and each member can also commit once (at least right now on master)

```suggestion
  , commits :: UTxO
```

ch1bo commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -0,0 +1,197 @@
+instance Arbitrary HeadStatus where
+  arbitrary = genericArbitrary
+
+data HeadState = HeadState
```
Shoud: add a haddock explaining why the status is only a simple tag and the data about the head is somethings partial (`Maybe`, `Unknown` etc)

ch1bo commented on hydra-explorer/json-schemas/hydra-explorer-api.yaml:
```diff
@@ -0,0 +1,195 @@
+    HeadState:
+      type: object
+      properties:
+        headId:
```
Should: `headId` should be `required` at least

ffakenz replied:
good catch!

ch1bo commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -0,0 +1,197 @@
+  , seedTxIn :: TxIn
+  , status :: HeadStatus
+  , contestationPeriod :: ContestationPeriod
+  , members :: [HeadMember]
```
Idea (not needed right away):

```suggestion
  , members :: [HeadMember]
  , contesters :: [Party]
```

ffakenz replied:
used `contestations :: Observed Natural` as agreed

ch1bo reviewed (changes requested):
Looks very good already! Needs some behavior changes on the explorer aggregation (with additional tests) and some minor cleanup.

ffakenz committed:
Improve README.md spelling

Co-authored-by: Sasha Bogicevic <Sasha.Bogicevic@iohk.io>

ffakenz committed:
Remove not needed endpoints

As we are not serving any files yet.

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
Remove unnecessary handle from withChainObserver and withHydraExplorer helpers

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
Update README

ffakenz committed:
Refactor e2e to use http simple library instead

As we do not need use http and to support TLS.

Also:
- moved the seedFromFaucet_ calls before starting a hydra-node.
- fixed the /heads endpoint header to use Accept.
- removed some unused packages after this refactor.
- remove e2e spec relying on stdout as we are only interested in the http api.

ffakenz committed:
Remove unimplemented endpoint from openapi spec

ffakenz committed:
Rename test module to match where the actual code is implemented

ffakenz committed:
Improve explorer spec to only assert on headers that we care about

ffakenz committed:
Update explorer api spec to specify required fields

ffakenz committed:
Use eitherDecode to print why the responded bytes where not a proper JSON value

ffakenz committed:
Refactor HydraExplorerHandle as an easy access to the API (= a client)

And assert REST API behavior rather than std out logs.

ffakenz committed:
Remove unused import from chain observer e2e spec

ffakenz committed:
Replace PartyCommit by UTxO type

ffakenz committed:
Increase the time dealy to wait for the explorer http server to be listening on port

Because it was flaky as it was defined.

ffakenz committed:
Draft property spec for aggregate head observations

Added arbitrary instances for head observations.

ffakenz committed:
Update aggregate observation logic to fulfill the expected properties

Also:
- refactor HeadState.seedTxIn to be Observed.
- refactor HeadState.contestationPeriod to be Observed.
- refactor HeadMember.onChainId to be Observed.

Finally updated the explorer openapi to be compliant with above changes.

ffakenz committed:
Refactor aggregation to work over OnChainTx type instead of HeadObservation

ffakenz committed:
Add tracking of contestation deadline as part of explorer's head states

ffakenz committed:
Add head state haddock

ffakenz committed:
Remove unsupported attribute propertyNames by swagger

ffakenz committed:
Minor fix after rebasing

v0d1ch committed:
Use 5 ADA for internal wallets

v0d1ch committed:
Remove misplaced comment

ffakenz committed:
Refactor http server to use servant

ffakenz committed:
Fix explorer e2e specs by increasing the amount of fuel provided to each party member

ffakenz committed:
Minor formatting changes

ch1bo commented on hydra-cluster/hydra-cluster.cabal:
```diff
@@ -168,6 +169,8 @@ test-suite tests
     , directory
     , filepath
     , hspec
+    , http-client
```
Could: remove this as http-conduit includes the same things

ch1bo commented on hydra-cluster/test/Test/HydraExplorerSpec.hs:
```diff
@@ -0,0 +1,134 @@
+              allHeads ^. nth 1 . key "headId" . _String `shouldBe` bobHeadId
+              allHeads ^. nth 1 . key "status" . _String `shouldBe` "Aborted"
+
+newtype HydraExplorerHandle = HydraExplorerHandle {getHeads :: IO Value}
```
:+1:

ch1bo commented on hydra-explorer/hydra-explorer.cabal:
```diff
@@ -0,0 +1,97 @@
+cabal-version: 3.0
+name:          hydra-explorer
+version:       0.15.0
```
Could: use latest hydra version (as long as the explorer is in the same repo / tagged together with `hydra-node`)

```suggestion
version:       0.16.0
```

ch1bo commented on hydra-explorer/json-schemas/hydra-explorer-api.yaml:
```diff
@@ -0,0 +1,206 @@
+openapi: 3.0.0
+info:
+  title: Head Explorer API
+  version: 1.0.0
```
Could: use latest hydra version (as long as the explorer is in the same repo / tagged together with `hydra-node`)

```suggestion
  version: 0.16.0
```

ch1bo commented on hydra-explorer/src/Hydra/Explorer.hs:
```diff
@@ -0,0 +1,116 @@
+             , Header "Access-Control-Allow-Origin" String
+             , Header "Access-Control-Allow-Methods" String
+             , Header "Access-Control-Allow-Headers" String
+             ]
```
Should: not add explicit `Headers`

The `'[JSON]` is already doing the content-type encoding including the request `Accept` header for you.

The CORS headers (the other three) we should add "around" the `API` when you `serve`, i.e. something like `serve (addCorsHeaders api)` or simply doing it via a middleware https://hackage.haskell.org/package/wai-cors-0.2.7/docs/Network-Wai-Middleware-Cors.html#v:simpleCors

ch1bo commented on hydra-explorer/src/Hydra/Explorer.hs:
```diff
@@ -0,0 +1,116 @@
+        [HeadState]
+    )
+handleGetHeads getHeads = do
+  result <- liftIO $ try getHeads
+  case result of
+    Right heads -> do
+      return $ addHeader "application/json" $ addCorsHeaders heads
+    Left (_ :: SomeException) -> throwError err500
```
Should: not use `try`, but have servant/warp handle the exceptions

In fact, the default handler will result in exactly the same thing as you have here. If you want more control, use [servant-exceptions](https://github.com/ch1bo/servant-exceptions/tree/master) (I will make it compatible to new versions of `servant` if you want to use it :smiley:)

```suggestion
  getHeads
```

ch1bo commented on hydra-explorer/src/Hydra/Explorer.hs:
```diff
@@ -0,0 +1,116 @@
+    Left (_ :: SomeException) -> throwError err500
+
+logMiddleware :: Tracer IO APIServerLog -> Middleware
+logMiddleware tracer app' req sendResponse = do
```
:+1:

ch1bo commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -0,0 +1,292 @@
+instance Arbitrary HeadState where
+  arbitrary = genericArbitrary
+
+type ExplorerState = [HeadState]
```
Could: move this to the top of the module (as it is the name-sake)

ch1bo commented on hydra-explorer/src/Hydra/Explorer/ExplorerState.hs:
```diff
@@ -0,0 +1,292 @@
+    (currentHeadState@HeadState{headId = currentHeadStateId} : tailStates) ->
+      if newHeadStateId == currentHeadStateId
+        then newHeadState : tailStates
+        else currentHeadState : replaceHeadState newHeadState tailStates
```
Should: use a better data type than a list.

How about a `Map HeadId HeadState`? Then we could replace this code with simply a `Map.insert` instead.

Or an insert ordered map (thinking about our properties): https://hackage.haskell.org/package/insert-ordered-containers-0.2.5.3/docs/Data-HashMap-Strict-InsOrd.html#t:InsOrdHashMap

ch1bo replied:
As we often have a `case findHeadState`, we could also use `Map.alter` to add/update elements also more easily.

ch1bo commented on hydra-explorer/test/Hydra/Explorer/ExplorerStateSpec.hs:
```diff
@@ -0,0 +1,30 @@
+
+spec :: Spec
+spec = do
+  describe "aggregate head observation into explorer state" $ do
```
```suggestion
  describe "aggregateHeadObservations" $ do
```

ch1bo commented on hydra-explorer/test/Hydra/Explorer/ExplorerStateSpec.hs:
:+1:

ch1bo reviewed (approved):
This looks nice now with servant!

Left some improvement suggestions as some of the code is a bit redundant.
